# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
#load python packages
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

In [2]:
os.getcwd()

'C:\\Users\\Umesh\\Documents\\GitHub\\SB_DS\\Career_Track\\Guided_Capstone\\Modeling'

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [3]:
df = pd.read_csv("../EDA/data/step3_output.csv")
df = df.drop("Unnamed: 0", axis=1)
df.head(5)

Name       state  summit_elev  vertical_drop  \
0               Hilltop Ski Area      Alaska         2090            294   
1            Sunrise Park Resort     Arizona        11100           1800   
2  Yosemite Ski & Snowboard Area  California         7800            600   
3         Boreal Mountain Resort  California         7700            500   
4               Donner Ski Ranch  California         8012            750   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0      0        0.0          0          0     0       1  ...   
1      0        0.0          0          1     2       3  ...   
2      0        0.0          0          0     0       1  ...   
3      0        0.0          0          1     1       3  ...   
4      0        0.0          0          0     0       1  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0            30.0             150.0       36.0   
1              800.0            80.0             115.0       49.0   
2               88.0            88.0             110.0       84.0   
3              380.0           200.0             150.0       54.0   
4              505.0            60.0             163.0       82.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0          30.0      34.00000              152.0   
1            250.0          74.0      78.00000              104.0   
2            300.0          47.0      47.00000              107.0   
3            400.0          49.0      64.16681              150.0   
4            400.0          75.0      75.00000              170.0   

   NightSkiing_ac  clusters  
0            30.0         1  
1            80.0         2  
2             0.0         2  
3           200.0         2  
4             0.0         2  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [4]:
df = pd.concat([df.drop('state', axis=1), pd.get_dummies(df['state'])], axis=1)
df.head(5)

Name  summit_elev  vertical_drop  trams  \
0               Hilltop Ski Area         2090            294      0   
1            Sunrise Park Resort        11100           1800      0   
2  Yosemite Ski & Snowboard Area         7800            600      0   
3         Boreal Mountain Resort         7700            500      0   
4               Donner Ski Ranch         8012            750      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0        0.0          0          0     0       1       0  ...             0   
1        0.0          0          1     2       3       1  ...             0   
2        0.0          0          0     0       1       3  ...             0   
3        0.0          0          1     1       3       1  ...             0   
4        0.0          0          0     0       1       5  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [5]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing as ppg

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = ppg.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

C:\Users\Umesh\Anaconda3\lib\site-packages\sklearn\preprocessing\data.py:645: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
C:\Users\Umesh\Anaconda3\lib\site-packages\ipykernel_launcher.py:14: DataConversionWarning: Data with input dtype uint8, int64, float64 were all converted to float64 by StandardScaler.
  


## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [6]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [7]:
X_train = X_train.astype(float)
y_train = y_train.astype(float)

In [8]:
np.any(np.isnan(X_train))

True

In [9]:
X_train[np.isnan(X_train)]=0

In [10]:
np.any(np.isnan(X_train))

False

In [11]:
np.all(np.isfinite(X_train))

True

In [12]:
from numpy import inf
X_train[X_train == -inf] = 0
X_train[X_train == +inf] = 0

In [13]:
np.all(np.isfinite(X_train))

True

In [14]:
np.any(np.isnan(y_train))

False

In [15]:
np.all(np.isfinite(y_train))

True

In [16]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [17]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [18]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [19]:
from sklearn.metrics import explained_variance_score as evs
from sklearn.metrics import mean_absolute_error as mae

print("Explained Variance Score = " + str(evs(y_test, y_pred)))
print("Mean Absolute Error = " + str(mae(y_test, y_pred)))

Explained Variance Score = 0.6166551831666583
Mean Absolute Error = 8.55713591208579


**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [20]:
model.intercept_

56.455459998230545

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [21]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

Coefficient
summit_elev        2.974328e+00
vertical_drop      3.623026e+00
trams              2.072901e+13
fastEight          1.985356e+13
fastSixes          1.425177e+13
fastQuads          2.661885e+12
quad               3.273243e+12
triple             5.663334e+12
double             6.554475e+12
surface            6.613179e+12
total_chairs       1.242911e+13
Runs               2.805664e+00
TerrainParks       6.289062e-01
LongestRun_mi      2.988281e-01
SkiableTerrain_ac  6.142578e-01
Snow Making_ac     8.125000e-01
daysOpenLastYear   1.924805e+00
yearsOpen          2.832031e-02
averageSnowfall    6.347656e-02
AdultWeekday       8.883789e+00
projectedDaysOpen  1.237305e+00
NightSkiing_ac     1.281250e+00
clusters           4.755859e-01
Alaska             2.725961e+11
Arizona            2.725961e+11
California         8.378885e+11
Colorado           4.692447e+11
Connecticut        6.020179e+11
Idaho              5.401518e+11
Illinois           4.692447e+11
Indiana            3.843247e+11
Iowa               4.692447e+11
Maine              6.020179e+11
Maryland           2.725961e+11
Massachusetts      5.401518e+11
Michigan           1.009403e+12
Minnesota          7.542809e+11
Missouri           3.843247e+11
Montana            3.843247e+11
Nevada             3.843247e+11
New Hampshire      8.759270e+11
New Jersey         2.725961e+11
New Mexico         6.020179e+11
New York           1.145835e+12
North Carolina     4.692447e+11
Ohio               5.401518e+11
Oregon             3.843247e+11
Pennsylvania       8.759270e+11
Rhode Island       2.725961e+11
South Dakota       2.725961e+11
Tennessee          2.725961e+11
Utah               4.692447e+11
Vermont            7.542809e+11
Virginia           3.843247e+11
Washington         3.843247e+11
West Virginia      4.692447e+11
Wisconsin          8.378885e+11
Wyoming            6.020179e+11

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [22]:
X.columns

Index(['summit_elev', 'vertical_drop', 'trams', 'fastEight', 'fastSixes',
       'fastQuads', 'quad', 'triple', 'double', 'surface', 'total_chairs',
       'Runs', 'TerrainParks', 'LongestRun_mi', 'SkiableTerrain_ac',
       'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen', 'averageSnowfall',
       'AdultWeekday', 'projectedDaysOpen', 'NightSkiing_ac', 'clusters',
       'Alaska', 'Arizona', 'California', 'Colorado', 'Connecticut', 'Idaho',
       'Illinois', 'Indiana', 'Iowa', 'Maine', 'Maryland', 'Massachusetts',
       'Michigan', 'Minnesota', 'Missouri', 'Montana', 'Nevada',
       'New Hampshire', 'New Jersey', 'New Mexico', 'New York',
       'North Carolina', 'Ohio', 'Oregon', 'Pennsylvania', 'Rhode Island',
       'South Dakota', 'Tennessee', 'Utah', 'Vermont', 'Virginia',
       'Washington', 'West Virginia', 'Wisconsin', 'Wyoming'],
      dtype='object')

In [23]:
X.shape

(164, 58)

In [24]:
X_no_state = X.loc[:, :'clusters']
X_no_state.head(5)

summit_elev  vertical_drop  trams  fastEight  fastSixes  fastQuads  quad  \
0         2090            294      0        0.0          0          0     0   
1        11100           1800      0        0.0          0          1     2   
2         7800            600      0        0.0          0          0     0   
3         7700            500      0        0.0          0          1     1   
4         8012            750      0        0.0          0          0     0   

   triple  double  surface  ...  LongestRun_mi  SkiableTerrain_ac  \
0       1       0        2  ...            1.0               30.0   
1       3       1        0  ...            1.2              800.0   
2       1       3        1  ...            0.4               88.0   
3       3       1        2  ...            1.0              380.0   
4       1       5        2  ...            1.5              505.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0            30.0             150.0       36.0             69.0          30.0   
1            80.0             115.0       49.0            250.0          74.0   
2            88.0             110.0       84.0            300.0          47.0   
3           200.0             150.0       54.0            400.0          49.0   
4            60.0             163.0       82.0            400.0          75.0   

   projectedDaysOpen  NightSkiing_ac  clusters  
0              152.0            30.0         1  
1              104.0            80.0         2  
2              107.0             0.0         2  
3              150.0           200.0         2  
4              170.0             0.0         2  

[5 rows x 23 columns]

In [25]:
X_no_state.columns

Index(['summit_elev', 'vertical_drop', 'trams', 'fastEight', 'fastSixes',
       'fastQuads', 'quad', 'triple', 'double', 'surface', 'total_chairs',
       'Runs', 'TerrainParks', 'LongestRun_mi', 'SkiableTerrain_ac',
       'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen', 'averageSnowfall',
       'AdultWeekday', 'projectedDaysOpen', 'NightSkiing_ac', 'clusters'],
      dtype='object')

In [26]:
X_no_state.shape

(164, 23)

In [27]:
# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler_no_state = ppg.StandardScaler().fit(X_no_state)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_no_state_scaled=scaler_no_state.transform(X_no_state) 

C:\Users\Umesh\Anaconda3\lib\site-packages\sklearn\preprocessing\data.py:645: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
C:\Users\Umesh\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  """


In [28]:
# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_no_state_train, X_no_state_test, y_train, y_test = train_test_split(X_no_state_scaled, y, test_size=0.25, random_state=1)

In [29]:
X_no_state_train[np.isnan(X_no_state_train)]=0
X_no_state_train[X_no_state_train == -inf] = 0
X_no_state_train[X_no_state_train == +inf] = 0

In [30]:
model_no_state = lm.fit(X_no_state_train,y_train)
model_no_state.intercept_

56.83445492594428

In [31]:
y_no_state_pred = model.predict(X_no_state_test)
print("Explained Variance Score = " + str(evs(y_test, y_no_state_pred)))
print("Mean Absolute Error = " + str(mae(y_test, y_no_state_pred)))

Explained Variance Score = 0.7459741649161075
Mean Absolute Error = 6.553948224752127


In [32]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X_no_state.columns, columns=['Coefficient'])

Coefficient
summit_elev        2.165995e+00
vertical_drop      2.630426e-01
trams              8.881784e-16
fastEight          7.771561e-16
fastSixes          7.216450e-16
fastQuads          5.481522e-01
quad               7.708505e-01
triple             1.336738e+00
double             9.534893e-01
surface            1.189149e+00
total_chairs       7.995952e-01
Runs               1.920583e+00
TerrainParks       1.286597e-01
LongestRun_mi      1.769247e-01
SkiableTerrain_ac  2.441872e-01
Snow Making_ac     1.387295e+00
daysOpenLastYear   1.638325e+00
yearsOpen          4.988187e-01
averageSnowfall    1.683611e+00
AdultWeekday       9.795630e+00
projectedDaysOpen  6.619612e-01
NightSkiing_ac     1.030786e+00
clusters           3.673978e+00

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [33]:
X_no_ssv = X.loc[:, 'trams':'clusters']
X_no_ssv.columns

Index(['trams', 'fastEight', 'fastSixes', 'fastQuads', 'quad', 'triple',
       'double', 'surface', 'total_chairs', 'Runs', 'TerrainParks',
       'LongestRun_mi', 'SkiableTerrain_ac', 'Snow Making_ac',
       'daysOpenLastYear', 'yearsOpen', 'averageSnowfall', 'AdultWeekday',
       'projectedDaysOpen', 'NightSkiing_ac', 'clusters'],
      dtype='object')

In [34]:
X_no_ssv.shape

(164, 21)

In [35]:
# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler_no_ssv = ppg.StandardScaler().fit(X_no_ssv)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_no_ssv_scaled=scaler_no_ssv.transform(X_no_ssv) 

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_no_ssv_train, X_no_ssv_test, y_train, y_test = train_test_split(X_no_ssv_scaled, y, test_size=0.25, random_state=1)

X_no_ssv_train[np.isnan(X_no_ssv_train)]=0
X_no_ssv_train[X_no_ssv_train == -inf] = 0
X_no_ssv_train[X_no_ssv_train == +inf] = 0

model_no_ssv = lm.fit(X_no_ssv_train,y_train)
model_no_ssv.intercept_

C:\Users\Umesh\Anaconda3\lib\site-packages\sklearn\preprocessing\data.py:645: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
C:\Users\Umesh\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: DataConversionWarning: Data with input dtype int64, float64 were all converted to float64 by StandardScaler.
  """


56.92520947786605

In [36]:
y_no_ssv_pred = model_no_ssv.predict(X_no_ssv_test)
print("Explained Variance Score = " + str(evs(y_test, y_no_ssv_pred)))
print("Mean Absolute Error = " + str(mae(y_test, y_no_ssv_pred)))

Explained Variance Score = 0.7584894253464819
Mean Absolute Error = 6.3897770848327164


In [37]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X_no_ssv.columns, columns=['Coefficient'])

Coefficient
trams              0.000000e+00
fastEight          0.000000e+00
fastSixes          1.776357e-15
fastQuads          6.877366e-01
quad               1.008003e+00
triple             1.416965e+00
double             8.158082e-01
surface            1.473275e+00
total_chairs       7.047187e-01
Runs               1.865500e+00
TerrainParks       3.022749e-01
LongestRun_mi      4.749233e-01
SkiableTerrain_ac  6.758046e-02
Snow Making_ac     9.648650e-01
daysOpenLastYear   2.210391e+00
yearsOpen          4.631642e-01
averageSnowfall    1.142104e+00
AdultWeekday       1.012469e+01
projectedDaysOpen  1.127213e+00
NightSkiing_ac     6.330989e-01
clusters           2.642517e+00

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.6166551831666583 | 8.55713591208579 |'base_elev'|
| Model 2. | 0.7459741649161075 | 6.553948224752127 |'state', 'base_elev'|
| Model 3. | 0.7584894253464819 | 6.3897770848327164 |'state','summit_elev','base_elev', 'vertical drop'|

Model Selection:

   ### Model 3 is my choice